six/modules/c++/six/source/NITFReadControl.cpp

```diff
@@ -127,34 +120,21 @@ DataType NITFReadControl::getDataType(nitf::DESegment& segment)
     }
     else if (desid == Constants::DES_USER_DEFINED_SUBHEADER_TAG)
     {
-        // Check whether DES is SICD/SIDD, or something of a different
-        // sort. Note that we need to check the subheader
-        // length first rather than calling tre.getCurrentSize() in
-        // case there is no subheader (in which case
-        // subheader.getSubheaderFields() will throw with a
-        // NITRO-specific message).
-        const sys::Uint64_T
-            subheaderLength(subheader.getSubheaderFieldsLength());
+        // Check whether DES is SICD/SIDD, or something of a different sort.
         if (subheaderLength !=
             Constants::DES_USER_DEFINED_SUBHEADER_LENGTH)
         {
             return DataType::NOT_SET;
         }
-        nitf::TRE tre = subheader.getSubheaderFields();
-        if (tre.getTag() !=
-            Constants::DES_USER_DEFINED_SUBHEADER_TAG)
+        if (treTag != Constants::DES_USER_DEFINED_SUBHEADER_TAG)
         {
             return DataType::NOT_SET;
         }
-
-        std::string field = tre.getField("DESSHSI").toString();
-        str::trim(field);
-
-        if (field == Constants::SICD_DESSHSI)
+        if (desshsiField == Constants::SICD_DESSHSI)
         {
             return DataType::COMPLEX;
         }
-        else if (field == Constants::SIDD_DESSHSI)
+        else if (desshsiField == Constants::SIDD_DESSHSI)
         {
             return DataType::DERIVED;
```

Comment:
Right now the function won't return anything if the conditions on lines 113, 117, and 121 are all false.  So, instead of the final else that returns `NOT_SET` on line 143, that should move all the way out and be the default return statement.